File: flamingo2-web/src/main/java/org/opencloudengine/flamingo2/engine/backend/UserEvent.java
/**
 * Copyright (C) 2011 Flamingo Project (http://www.cloudine.io).
 * <p/>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p/>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p/>
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.opencloudengine.flamingo2.engine.backend;

import org.opencloudengine.flamingo2.core.security.SessionUtils;
import org.opencloudengine.flamingo2.util.DateUtils;
import org.opencloudengine.flamingo2.util.JVMIDUtils;

import java.io.Serializable;
import java.sql.Timestamp;
import java.util.Date;

public class UserEvent implements Serializable {

    Long id;
    String name;
    Timestamp registrationDate;
    String status;
    String message;
    Boolean isSee;
    String identifier;
    Long referenceId;
    String username;
    String yyyy;
    String mm;
    String dd;

    public static UserEvent create(String username, String name, String status) {
        UserEvent event = new UserEvent();
        event.name = name;
        event.isSee = false;
        event.status = status;
        event.username = username;
        Date date = new Date();
        event.registrationDate = new Timestamp(date.getTime());
        event.yyyy = DateUtils.parseDate(date, "yyyy");
        event.mm = DateUtils.parseDate(date, "MM");
        event.dd = DateUtils.parseDate(date, "dd");

        event.identifier = event.yyyy + event.mm + event.dd + "_" + System.currentTimeMillis() + "_" + JVMIDUtils.generateUUID();
        return event;
    }

    public static UserEvent create(String name, String status) {
        UserEvent event = new UserEvent();
        event.name = name;
        event.isSee = false;
        event.status = status;
        event.username = SessionUtils.getUsername();
        Date date = new Date();
        event.registrationDate = new Timestamp(date.getTime());
        event.yyyy = DateUtils.parseDate(date, "yyyy");
        event.mm = DateUtils.parseDate(date, "MM");
        event.dd = DateUtils.parseDate(date, "dd");

        event.identifier = event.yyyy + event.mm + event.dd + "_" + System.currentTimeMillis() + "_" + JVMIDUtils.generateUUID();
        return event;
    }

    public static UserEvent create(String name, String status, String message, String identifier, Long referenceId) {
        UserEvent event = new UserEvent();
        event.name = name;
        event.isSee = false;
        event.identifier = identifier;
        event.referenceId = referenceId;
        event.status = status;
        event.message = message;
        event.username = SessionUtils.getUsername();

        Date date = new Date();
        event.registrationDate = new Timestamp(date.getTime());
        event.yyyy = DateUtils.parseDate(date, "yyyy");
        event.mm = DateUtils.parseDate(date, "MM");
        event.dd = DateUtils.parseDate(date, "dd");
        return event;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Timestamp getRegistrationDate() {
        return registrationDate;
    }

    public void setRegistrationDate(Timestamp registrationDate) {
        this.registrationDate = registrationDate;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public Boolean getIsSee() {
        return isSee;
    }

    public void setIsSee(Boolean isSee) {
        this.isSee = isSee;
    }

    public String getIdentifier() {
        return identifier;
    }

    public void setIdentifier(String identifier) {
        this.identifier = identifier;
    }

    public Long getReferenceId() {
        return referenceId;
    }

    public void setReferenceId(Long referenceId) {
        this.referenceId = referenceId;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getYyyy() {
        return yyyy;
    }

    public void setYyyy(String yyyy) {
        this.yyyy = yyyy;
    }

    public String getMm() {
        return mm;
    }

    public void setMm(String mm) {
        this.mm = mm;
    }

    public String getDd() {
        return dd;
    }

    public void setDd(String dd) {
        this.dd = dd;
    }

    @Override
    public String toString() {
        return "ID='" + id + '\'' +
                ", Name='" + name + '\'' +
                ", Registration=" + registrationDate +
                ", Status='" + status + '\'' +
                ", Message='" + message + '\'' +
                ", IsSee='" + isSee + '\'' +
                ", Identifier='" + identifier + '\'' +
                ", ReferenceId='" + referenceId + '\'' +
                ", Username='" + username + '\'' +
                ", YYYY='" + yyyy + '\'' +
                ", MM='" + mm + '\'' +
                ", DD='" + dd + '\'';
    }
}

File: flamingo2-web/src/main/webapp/resources/app/view/admin/user/UserEvent.js
/*
 * Copyright (C) 2015 Bahamas Project (http://www.cloudine.io).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
Ext.define('Flamingo2.view.admin.user.UserEvent', {
    extend: 'Flamingo2.panel.Panel',
    alias: 'widget.userevent',

    requires: [
        'Flamingo2.view.admin.user.UserEventController',
        'Flamingo2.view.admin.share.Model'
    ],
    viewModel: {
        type: 'admin.shareModel'
    },

    controller: 'userEventController',

    layout: {
        type: 'vbox',
        align: 'stretch'
    },
    items: [
        {
            xtype: 'grid',
            itemId: 'usereventGrid',
            reference: 'usereventGrid',
            border: 0,
            viewConfig: {
                stripeRows: true,
                columnLines: true,
                enableTextSelection: true,
                getRowClass: function (b, e, d, c) {
                    return 'cell-height-30';
                }
            },
            bind: {
                store: '{userEvent}'
            },
            columns: [
                {text: message.msg('batch.job_name'), dataIndex: 'name', flex: 1},
                {text: message.msg('common.status'), dataIndex: 'status', width: 120, align: 'center'},
                {
                    text: message.msg('common.detail'), width: 100, align: 'center',
                    renderer: function (val, meta, rec) {
                        var id = Ext.id();
                        Ext.defer(function () {
                            Ext.create('Ext.button.Button', {
                                renderTo: id,
                                text: message.msg('common.view'),
                                iconCls: 'common-view',
                                usage: 'vmproperty-apply',
                                scale: 'small',
                                handler: function () {
                                    var popup = Ext.create('Ext.Window', {
                                        title: message.msg('common.history'),
                                        width: 850,
                                        height: 500,
                                        layout: 'fit',
                                        maximizable: true,
                                        items: {
                                            xtype: 'workflowDetail',
                                            identifier: rec.data.identifier
                                        }
                                    }).center().show();
                                }
                            });
                        }, 50);
                        return Ext.String.format('<div id="{0}"></div>', id);
                    }
                },
                {
                    text: message.msg('common.select'),
                    width: 50,
                    align: 'center',
                    xtype: 'checkcolumn',
                    dataIndex: 'isSee'
                }

            ],
            dockedItems: [
                {
                    xtype: 'pagingtoolbar',
                    bind: {
                        store: '{userEvent}'
                    },
                    dock: 'bottom'
                }
            ],
            listeners: {
                itemclick: function (view, record, item, index, e) {
                }
            }
        }
    ],
    tbar: [
        {
            xtype: 'tbtext',
            text: message.msg('hdfs.audit.tbar.chart.condition')
        },
        {
            xtype: 'combo',
            width: 130,
            itemId: 'conditionKey',
            name: 'conditionKey',
            displayField: 'name',
            valueField: 'value',
            emptyText: '- ' + message.msg('common.select') + ' -',
            editable: true,
            selectOnFocus: true,
            collapsible: false,
            store: Ext.create('Ext.data.Store', {
                fields: ['name', 'value'],
                data: [
                    {name: message.msg('batch.job_name'), value: 'NAME'}
                ]
            }),
            listeners: {
                change: function (combo, newValue, oldValue, eOpts) {
                    var store = query('userevent #usereventGrid').getStore();
                    store.getProxy().extraParams.conditionKey = newValue;
                }
            }
        },
        {
            xtype: 'textfield',
            width: 130,
            emptyText: '',
            name: 'conditionValue',
            itemId: 'conditionValue'
        },
        {
            xtype: 'button',
            itemId: 'findButton',
            formBind: true,
            text: message.msg('common.retrieve'),
            iconCls: 'common-find',
            labelWidth: 50,
            handler: function (button) {
                var store = query('userevent #usereventGrid').getStore();
                var conditionKey = store.getProxy().extraParams.conditionKey;
                var conditionValue = query('userevent #conditionValue').getValue().trim();
                store.getProxy().extraParams.conditionValue = conditionValue;

                if (conditionKey && conditionValue) {
                    store.getProxy().extraParams = {
                        conditionKey: conditionKey,
                        conditionValue: conditionValue
                    };
                    store.loadPage(1);
                } else if (!conditionKey || !conditionValue) {
                    store.getProxy().extraParams = {};
                    store.loadPage(1);
                } else {
                    Ext.MessageBox.show({
                        title: message.msg('common.error'),
                        message: message.msg('common.msg.enter_condition'),
                        buttons: Ext.MessageBox.OK,
                        icon: Ext.MessageBox.WARNING,
                        fn: function (e) {
                            return false;
                        }
                    });
                }
            }
        },
        {
            xtype: 'button',
            itemId: 'clearButton',
            formBind: true,
            text: message.msg('system.user.common.clear'),
            iconCls: 'common-find-clear',
            labelWidth: 50,
            handler: function () {
                var store = query('userevent #usereventGrid').getStore();
                store.getProxy().extraParams = {};
                query('userevent #conditionKey').clearValue();
                query('userevent #conditionValue').setValue('');
            }
        },
        {
            xtype: 'button',
            itemId: 'deleteButton',
            formBind: true,
            text: message.msg('common.delete'),
            iconCls: 'common-delete',
            labelWidth: 50,
            handler: function () {
                var identifiers = [];
                var store = query('userevent #usereventGrid').getStore();
                Ext.each(store.data.items, function (item) {
                    if (item.data.isSee) {
                        identifiers.push(item.data.identifier);
                    }
                });
                if (identifiers.length > 0) {
                    invokePostByMap(CONSTANTS.USER.PREFERENCE.EVENT.SAW,
                        {identifiers: identifiers},
                        function (response) {
                            store.load();
                        },
                        function (response) {
                            store.load();
                        }
                    );
                }
            }
        }
    ],

    listeners: {
        afterrender: function () {
            var store = this.getViewModel().getStore('userEvent');
            store.getProxy().extraParams.conditionKey = '';
            store.getProxy().extraParams.conditionValue = '';
            store.load();
        }
    }

});